[Catalog] Fail to fetch TPU-v3-64 spot price on us-east1

```
  File "/opt/hostedtoolcache/Python/3.10.13/x64/lib/python3.10/site-packages/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py", line 465, in get_tpu_price
    assert tpu_price is not None, row
AssertionError: AcceleratorName      tpu-v3-64
AcceleratorCount             1
Region                us-east1
AvailabilityZone    us-east1-d
Price                     64.0
```

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -473,7 +473,7 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             if not hidden_tpu.empty:
                 price_str = 'SpotPrice' if spot else 'Price'
                 tpu_price = hidden_tpu[price_str].values[0]
-        assert tpu_price is not None, row
+        assert tpu_price is not None, (row, hidden_tpu, HIDDEN_TPU_DF)
         return tpu_price
 
     df['Price'] = df.apply(lambda row: get_tpu_price(row, spot=False), axis=1)

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -472,6 +472,10 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             if not hidden_tpu.empty:
                 price_str = 'SpotPrice' if spot else 'Price'
                 tpu_price = hidden_tpu[price_str].values[0]
+        if tpu_price is None:
+            spot_str = 'spot ' if spot else ''
+            print(f'The {spot_str}price of {tpu_name} in {tpu_region} is '
+                    'not found in SKUs or hidden TPU price DF.')
         assert spot or tpu_price is not None, (row, hidden_tpu, HIDDEN_TPU_DF)
         return tpu_price
 

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -473,7 +473,6 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             if not hidden_tpu.empty:
                 price_str = 'SpotPrice' if spot else 'Price'
                 tpu_price = hidden_tpu[price_str].values[0]
-        assert tpu_price is not None, (row, hidden_tpu, HIDDEN_TPU_DF)
         return tpu_price
 
     df['Price'] = df.apply(lambda row: get_tpu_price(row, spot=False), axis=1)

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -422,9 +422,20 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
         return df
 
     def get_tpu_price(row: pd.Series, spot: bool) -> float:
+        assert row['AcceleratorCount'] == 1, row
         tpu_price = None
+        tpu_region = row['Region']
+        tpu_name = row['AcceleratorName']
+        tpu_version = tpu_name.split('-')[1]
+        num_cores = int(tpu_name.split('-')[2])
+        # For TPU-v2 and TPU-v3, the pricing API provides the prices
+        # of 8 TPU cores. The prices can be different based on
+        # whether the TPU is a single device or a pod.
+        # For TPU-v4, the pricing is uniform, and thus the pricing API
+        # only provides the price of TPU-v4 pods.
+        is_pod = num_cores > 8 or tpu_version == 'v4'
+
         for sku in skus:
-            tpu_region = row['Region']
             if tpu_region not in sku['serviceRegions']:
                 continue
             description = sku['description']
@@ -436,15 +447,6 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
                 if 'Preemptible' in description:
                     continue
 
-            tpu_name = row['AcceleratorName']
-            tpu_version = tpu_name.split('-')[1]
-            num_cores = int(tpu_name.split('-')[2])
-            # For TPU-v2 and TPU-v3, the pricing API provides the prices
-            # of 8 TPU cores. The prices can be different based on
-            # whether the TPU is a single device or a pod.
-            # For TPU-v4, the pricing is uniform, and thus the pricing API
-            # only provides the price of TPU-v4 pods.
-            is_pod = num_cores > 8 or tpu_version == 'v4'
 
             if f'Tpu-{tpu_version}' not in description:
                 continue
@@ -459,16 +461,24 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             tpu_device_price = unit_price
             tpu_core_price = tpu_device_price / 8
             tpu_price = num_cores * tpu_core_price
-            assert row['AcceleratorCount'] == 1, row
             break
-
+        
+        if tpu_price is None:
+            # Find the line with the same accelerator name, count and region in
+            # the hidden TPU dataframe for the row.
+            hidden_tpu = HIDDEN_TPU_DF[
+                (HIDDEN_TPU_DF['AcceleratorName'] == row['AcceleratorName']) &
+                (HIDDEN_TPU_DF['AcceleratorCount'] == row['AcceleratorCount']) &
+                (HIDDEN_TPU_DF['Region'] == row['Region'])]
+            if not hidden_tpu.empty:
+                price_str = 'SpotPrice' if spot else 'Price'
+                tpu_price = hidden_tpu[price_str].values[0]
         assert tpu_price is not None, row
         return tpu_price
 
     df['Price'] = df.apply(lambda row: get_tpu_price(row, spot=False), axis=1)
     df['SpotPrice'] = df.apply(lambda row: get_tpu_price(row, spot=True),
                                axis=1)
-    df = pd.concat([df, HIDDEN_TPU_DF], ignore_index=True)
     df = df.reset_index(drop=True)
     df['version_and_size'] = df['AcceleratorName'].apply(
         lambda name: (name.split('-')[1], int(name.split('-')[2])))

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -475,7 +475,7 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
         if tpu_price is None:
             spot_str = 'spot ' if spot else ''
             print(f'The {spot_str}price of {tpu_name} in {tpu_region} is '
-                    'not found in SKUs or hidden TPU price DF.')
+                  'not found in SKUs or hidden TPU price DF.')
         assert spot or tpu_price is not None, (row, hidden_tpu, HIDDEN_TPU_DF)
         return tpu_price
 

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -464,12 +464,12 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             break
         
         if tpu_price is None:
-            # Find the line with the same accelerator name, count and region in
+            # Find the line with the same accelerator name, region, zone in
             # the hidden TPU dataframe for the row.
             hidden_tpu = HIDDEN_TPU_DF[
                 (HIDDEN_TPU_DF['AcceleratorName'] == row['AcceleratorName']) &
-                (HIDDEN_TPU_DF['AcceleratorCount'] == row['AcceleratorCount']) &
-                (HIDDEN_TPU_DF['Region'] == row['Region'])]
+                (HIDDEN_TPU_DF['Region'] == row['Region']) &
+                (HIDDEN_TPU_DF['AvailabilityZone'] == row['AvailabilityZone'])]
             if not hidden_tpu.empty:
                 price_str = 'SpotPrice' if spot else 'Price'
                 tpu_price = hidden_tpu[price_str].values[0]

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -421,7 +421,7 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
     if df.empty:
         return df
 
-    def get_tpu_price(row: pd.Series, spot: bool) -> float:
+    def get_tpu_price(row: pd.Series, spot: bool) -> Optional[float]:
         assert row['AcceleratorCount'] == 1, row
         tpu_price = None
         tpu_region = row['Region']
@@ -447,7 +447,6 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
                 if 'Preemptible' in description:
                     continue
 
-
             if f'Tpu-{tpu_version}' not in description:
                 continue
             if is_pod:
@@ -462,7 +461,7 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             tpu_core_price = tpu_device_price / 8
             tpu_price = num_cores * tpu_core_price
             break
-        
+
         if tpu_price is None:
             # Find the line with the same accelerator name, region, zone in
             # the hidden TPU dataframe for the row.

--- a/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
+++ b/sky/clouds/service_catalog/data_fetchers/fetch_gcp.py
@@ -473,6 +473,7 @@ def get_tpu_df(skus: List[Dict[str, Any]]) -> pd.DataFrame:
             if not hidden_tpu.empty:
                 price_str = 'SpotPrice' if spot else 'Price'
                 tpu_price = hidden_tpu[price_str].values[0]
+        assert spot or tpu_price is not None, (row, hidden_tpu, HIDDEN_TPU_DF)
         return tpu_price
 
     df['Price'] = df.apply(lambda row: get_tpu_price(row, spot=False), axis=1)

